sdk/state/state_open.go

```diff
@@ -0,0 +1,151 @@
+	FormationSignatures   []xdr.DecoratedSignature
+}
+
+func (c *Channel) OpenTxs() (close, decl, formation *txnbuild.Transaction, err error) {
+	close, err = txbuild.Close(txbuild.CloseParams{
+		ObservationPeriodTime:      c.observationPeriodTime,
+		ObservationPeriodLedgerGap: c.observationPeriodLedgerGap,
+		InitiatorSigner:            c.initiatorSigner(),
+		ResponderSigner:            c.responderSigner(),
+		InitiatorEscrow:            c.initiatorEscrowAccount().Address,
+		ResponderEscrow:            c.responderEscrowAccount().Address,
+		StartSequence:              c.startingSequence,
+		IterationNumber:            1,
+		AmountToInitiator:          0,
+		AmountToResponder:          0,
+	})
+	if err != nil {
+		return
+	}
+	decl, err = txbuild.Declaration(txbuild.DeclarationParams{
+		InitiatorEscrow:         c.initiatorEscrowAccount().Address,
+		StartSequence:           c.startingSequence,
+		IterationNumber:         1,
+		IterationNumberExecuted: 0,
+	})
+	if err != nil {
+		return
+	}
+	formation, err = txbuild.Formation(txbuild.FormationParams{
+		InitiatorSigner: c.initiatorSigner(),
+		ResponderSigner: c.responderSigner(),
+		InitiatorEscrow: c.initiatorEscrowAccount().Address,
+		ResponderEscrow: c.responderEscrowAccount().Address,
+		StartSequence:   c.startingSequence,
+	})
+	return
+}
+
+// OpenPropose proposes the open of the channel, it is called by the participant
+// initiating the channel.
+func (c *Channel) OpenPropose() (Open, error) {
+	c.startingSequence = c.initiatorEscrowAccount().SequenceNumber + 1
+
+	close, _, _, err := c.OpenTxs()
+	if err != nil {
+		return Open{}, err
+	}
+	closeSig, err := c.sign(close)
+	if err != nil {
+		return Open{}, err
+	}
+	open := Open{
+		CloseSignatures: []xdr.DecoratedSignature{closeSig},
+	}
+	return open, nil
+}
+
+// OpenConfirm confirms an open that was proposed. It is called by both
+// participants as they both participate in the open process.
+//
+// If there are no sigs on the open, the open is invalid, it must be signed by
+// the proposing participant.
+//
+// If there is a close signature on the open by the remote participant, the
+// local participant will add close and/or declaration signatures, as required.
+//
+// If there are close and declaration signatures for all participants, the local
+// participant will add a formation signature, if required.
+//
+// If there are close, declaration, and formation signatures for all
+// participants, the channel will be considered open.
+func (c *Channel) OpenConfirm(m Open) (Open, error) {
+	c.startingSequence = c.initiatorEscrowAccount().SequenceNumber + 1
+
+	close, decl, formation, err := c.OpenTxs()
+	if err != nil {
+		return m, err
+	}
+
+	// If remote has not signed close, error as is invalid.
+	err = c.verifySigned(close, m.CloseSignatures, c.remoteSigner)
+	if err != nil {
+		return m, fmt.Errorf("open confirm: close invalid %w", err)
+	}
+
+	// If local has not signed close, sign it.
```

Comment:
I like doing it this way of checking if I've signed before signing, instead of the way I did. I'll change mine after we merge

Reply:
I also think we've used some different strategies around how we generate and handle errors we should make the same.